README

OPTICS : single-linkage point ordering
--------------------------------------
OPTICS is an algorithm to detect the single-linkage structure in data
collections. The algorithm detects, for each point, the number of 
neighbours within a pre-defined radius 'epsilon' and, if the number
is larger than the predefined threshold 'minPoints', proceeds with the
closest neighbour point, otherwise with the next point in the list of
unprocessed points.
OPTICS does not provide a clustering structure per se, but the
point ordering together with the closest-neighbour information can be
used to extract a cluster structure. The algorithm is useful for large
data sets with overlapping areas of varying point density.

The code has been tested to work for >300000 data points.

The algorithm requires only two input para-
meters: a neighbourhood radius (epsilon) and a minimum
number of neighbour points (MinPts). If we set epsilon to
the largest distance in the dataset then each point is reachable 
and this parameter has no influence on the results.
Briefly, the algorithm starts at a random data point, cal-
culates the distance to all points within the neighbourhood
radius (epsilon) and, if at least a minimal number of points
(MinPts) is encountered, it records the nearest neighbour
distance (Reachability Distance) and the smallest radius
that encircles MinPts objects (Core Distance). If less than
MinPts points fall within epsilon, the point is considered as
noise. The algorithm repeats the same procedure for the
nearest neighbour point and proceeds iteratively until all
data points have been visited, thereby generating an
ordered list. Setting epsilon to the largest distance implies that none of
the fragments is labelled 'noise' and each is included in
one cluster at least. This choice allows to scan afterwards
for clusters at any density.
The ordered list of Reachability Distances (RDs) can be drawn as 
a comprehensive nearest neighbour distance plot (called Reachability Plot).

The output files are as follows:
- file 'output.dat' contains the ordered list of input data;
  given are their location 'dataID' in the input data file, core distance 'CD' and 
  reachability distance 'RD'.
- file 'cluster.dat' contains the cluster structure of the data;
  given are the cluster identifier 'id', the id of the 'parent' cluster,
  'start' and 'end' in the ordered list and 'size' of the cluster,
  its smallest core distance 'minCD' and the identifier 'minCDid' of that point.
- file 'center.dat' contains the cluster centres;
  given are the centre's location 'index', identifier in the ordered list 'order',
  cluster identifier 'clusterID', core distance 'CD', 
  reachability distance 'RD' and coordinates 'x' 'y' 'z'. 


References
----------
Users publishing results obtained with the program should acknowledge its use by citation:
- Pandini, A, Fornili, A, Kleinjung, J
Structural alphabets derived from attractors in conformational space.
BMC Bioinformatics 11:97 (20 February), 2010.

Other references:
- Ankerst M, Breunig MM, Kriegel HP, Sander J: OPTICS: Ordering Points 
To Identify the Clustering Structure. 
SIGMOD Proceedings ACM SIGMOD International Conference on Management of Data, 
June 1-3, 1999, Philadelphia, Pennsylvania, USA ACM PressDelis A, Faloutsos C, 
Ghandeharizadeh S 1999, 49-60.

- Daszykowski M, Walczak B, Massart DL: Looking for natural patterns in analytical data. 
2. Tracing local density with OPTICS. 
Journal of chemical information and computer sciences 2002, 42(3):500-507.


Install / Uninstall
-------------------
Please read the general 'INSTALL' instructions.

There are several versions of the program that are invoked 
upon compilation with 'configure --enable-data-<version>':
- optics_ang
  for input data in angle coordinates
  compile using the command './configure --enable-data-ang'
- optics_str
  for input data in string format 
  compile using the command './configure --enable-data-str'
- optics_vec
  for input data in vector (of arbitrary length) format
  compile using the command './configure --enable-data-vec'
- optics_xyz 
  for input data in xyz (Cartesian) coordinates
  compile using the command './configure --enable-data-xyz'
- optics_dist 
  for input data as distance matrix
  compile using the command './configure --enable-data-dist'

The 'optics_ang' program requires the 'pcre' pattern matching library.
See http://www.pcre.org/ for installation options.

The 'optics_vec' program has several distance metrics, one of which is
enabled by a '#define' instruction at the top of the program.

Run 'configure --help' for more information.
For documentation execute 'doxygen doxygen.cfg' in the 'src' directory.
Documentation files are created in 'doc/html' and 'doc/latex'.
The latex documentation is completed by executing 'make pdf' in the
'doc/latex' directory, which creates 'refman.ps' and 'refman.pdf'.


Usage
-----
optics: perform OPTICS single-linkage point ordering
             (C) 2008-2015 by Jens Kleinjung and Alessandro Pandini
             This program is free software.

optics [--datafile ...] [OPTIONS ...]
      --datafile <filename>                (mode: mandatory, type: char,  default: input.dat)
      --outputfile <filename>              (mode: optional,  type: char,  default: output.dat)
      --clusterfile <filename>             (mode: optional,  type: char,  default: cluster.dat)
      --centerfile <filename>              (mode: optional,  type: char,  default: center.dat)
      --uniquefile <filename>              (mode: optional,  type: char,  default: unique.dat)
      --eps <epsilon cutoff>               (mode: optional,  type: float, default: 2.91)
      --minpts <min number of neighbours>  (mode: optional,  type: int,   default: 1)
      --w <window size for string version> (mode: optional,  type: int,   default: 1)
      --silent                             (mode: optional,  no argument, default: off)
      --cite                               (mode: optional , type: no_arg, default: off)
      --version                            (mode: optional , type: no_arg, default: off)
      --help

	Use the configure options described in 'README Install/Uninstall'
    to compile a specific verion.


Input formats
-------------
- optics_ang


- optics_str


- optics_vec
The input format is array of vectors of arbitrary length composed
of space separated floats.

- optics_xyz 


- optics_dist 


- optics_dist
The input format is a triangular matrix of distance values
in float format. See for example the 'tests/dist.dat' file.
The matrix contains the distances of the [i,j] point pairs
[0,1] [0,2] [0,3] [0,4] [0,5]
      [1,2] [1,3] [1,4] [1,5]
            [2,3] [2,4] [2,5]
                  [3,4] [3,5]
                        [4,5]
with values
5 3 2 8 1
  7 4 7 2
    4 2 4
      7 3
        2

White space and line breaks are irrelevant for the input format,
only the order of values in the above sense must be correct.


Return values
-------------
0 : Clean termination.
1 : Error.


-------------------------------------------------------------------------------
Copyright (C) 2008-2016 Jens Kleinjung
Copyright (C) 2008-2016 Alessandro Pandini

The development was kindly supported by grants from the
European Science Foundation "Frontiers of Functional Genomics" and
"Marie-Curie Intra-European Fellowships for Career development"
(Call identifier: PEOPLE-2007-2-1-IEF).
Further support was given by the MRC National Institute for Medical Research.

Availability
------------
The program is made available under the GNU Public License for academic
scientific purposes, under the condition that proper acknowledgement
is made to the authors of the program in publications resulting from the use
of the program.

License
-------
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.



ChangeLog
- 1.2.0
	Sat 17 Sep 13:50:42 BST 2016
	Modified vector version: several distance metrics can be set
	with #define.

- 1.1.5
	Tue Feb  4 13:49:01 GMT 2014
	Bug fix in loop optics.c:170-173: missing brackets in for loop

- 1.1.4
	Thu Feb 28 11:48:42 GMT 2013
	Added 'outpath' option

- 1.1.3
	Wed Feb 27 12:00:11 GMT 2013
	Added debugging and profiling options

- 1.1.2
	Tue Feb 19 00:16:48 GMT 2013
	Added n-dimensional Cartesian distance for vector comparison

- 1.1.1
	Wed Feb  6 10:55:16 GMT 2013
	Fixed two bugs related to input reading and normalisation check

- 1.1.0
	Thu Jan 31 13:14:36 GMT 2013
	Added option --enable-data-vec

- 1.0.1
	Sun Jul 29 13:25:53 BST 2012
	Changed function name 'mergesort' to 'mergesort_hlp' to avoid conflict
	with 'stdlib.h'.



src/arg.c
/*==============================================================================
arg.c : parse command line arguments
(C) 2008-2015 Jens Kleinjung and Alessandro Pandini
Read the COPYING file for license information.
==============================================================================*/

#include "arg.h"
#include "config.h"

/*____________________________________________________________________________*/
/** print version */
static void print_version()
{
	fprintf(stdout, "\n%s %s\n", PACKAGE, VERSION);
}

/*____________________________________________________________________________*/
/** print header */
static void print_header()
{
    fprintf(stdout, "\nOPTICS: single-linkage point ordering\n\n");
}

/*____________________________________________________________________________*/
/** print license */
static void print_license()
{
    fprintf(stderr, "\nCopyright (C) 2008-2016 by Jens Kleinjung and Alessandro Pandini\n"
					"You are welcome to redistribute it under certain conditions.\n"
                    "Read the COPYING file for license information.\n\n");
}

/*____________________________________________________________________________*/
/** print citation */
static void print_citation()
{
	printf("\n- Pandini et al., BMC Bioinformatics 11:97, 2010.\n");
	printf("- Ankerst et al., Proc.ACM SIGMOD'99 Int. Conf. on Management of Data, Philadelphia PA, 1999.\n");
	printf("- Daszykowski et al., J. Chem. Inf. Comput. Sci. 42:500-507, 2002.\n\n");
}

/*____________________________________________________________________________*/
/** set defaults */
static void set_defaults(Arg *arg)
{
    arg->dataInFileName = "input.dat";
    arg->dataOutFileName = "output.dat";
    arg->clusterOutFileName = "cluster.dat";
    arg->centerOutFileName = "center.dat";
    arg->uniqueOutFileName = "unique.dat";
    arg->eps = FLT_MAX;
    arg->minPts = 1;
    arg->w = 1;
    arg->outPathName = ".";
}

/*____________________________________________________________________________*/
/** parse command line long_options */
int parse_args(int argc, char **argv, Arg *arg)
{
    extern int silent;

    int c;
    const char usage[] = "\noptics [--datafile ...] [OPTIONS ...]\n\
      --datafile <filename>                (mode: mandatory, type: char,  default: input.dat)\n\
      --outpath <path>                     (mode: optional,  type: char,  default: .)\n\
      --outputfile <filename>              (mode: optional,  type: char,  default: output.dat)\n\
      --clusterfile <filename>             (mode: optional,  type: char,  default: cluster.dat)\n\
      --centerfile <filename>              (mode: optional,  type: char,  default: center.dat)\n\
      --uniquefile <filename>              (mode: optional,  type: char,  default: unique.dat)\n\
      --eps <epsilon cutoff>               (mode: optional,  type: float, default: FLT_MAX)\n\
      --minpts <min number of neighbours>  (mode: optional,  type: int,   default: 1)\n\
      --w <window size for string version> (mode: optional,  type: int,   default: 1)\n\
      --silent                             (mode: optional,  no argument, default: off)\n\
      --cite                               (mode: optional , type: no_arg, default: off)\n\
      --version                            (mode: optional , type: no_arg, default: off)\n\
      --help\n\n\
    optics_ang : input data in angle coordinates\n\
    optics_str : input data in string format\n\
    optics_vec : input data in vector (of arbitrary length) coordinates\n\
    optics_xyz : input data in Euclidean xyz coordinates\n\n\
    Use the configure options described in 'README' to compile a specific verion.\n";

    set_defaults(arg);

    if (argc < 2) {
		print_header();
        fprintf(stderr, "%s", usage);
		print_license();
        exit(1);
    }

    /** long option definition */
    static struct option long_options[] =
    {
        {"datafile", required_argument, 0, 1},
        {"outpath", required_argument, 0, 10},
        {"outputfile", required_argument, 0, 2},
        {"clusterfile", required_argument, 0, 3},
        {"centerfile", required_argument, 0, 4},
        {"uniquefile", required_argument, 0, 5},
        {"eps", required_argument, 0, 6},
        {"minpts", required_argument, 0, 7},
        {"w", required_argument, 0, 8},
        {"silent", no_argument, 0, 9},
        {"cite", no_argument, 0, 22},
        {"version", no_argument, 0, 23},
        {"help", no_argument, 0, 24},
        {0, 0, 0, 0}
    };

    /** assign parameters to long options */
    while ((c = getopt_long(argc, argv, "1:2:3:4:5:6:7:8:9 10: 22 23 24", long_options, NULL)) != -1)
    {
        switch(c)
        {
            case 1:
                arg->dataInFileName = optarg;
                break;
            case 2:
                arg->dataOutFileName = optarg;
                break;
            case 3:
                arg->clusterOutFileName = optarg;
                break;
            case 4:
                arg->centerOutFileName = optarg;
                break;
            case 5:
                arg->uniqueOutFileName = optarg;
                break;
            case 6:
                arg->eps = atof(optarg);
                break;
            case 7:
                arg->minPts = atoi(optarg);
                break;
            case 8:
                arg->w = atoi(optarg);
                break;
            case 9:
                ++ silent;
                break;
            case 10:
                arg->outPathName = optarg;
                break;
            case 22:
                print_citation();
                exit(0);
            case 23:
				print_version();
				print_license();
                exit(0);
            case 24:
                fprintf(stderr, "%s", usage);
				print_license();
                exit(0);
            default:
                fprintf(stderr, "%s", usage);
				print_license();
                exit(1);
        }
    }

    if (! silent) {
		print_header();
		fflush(stdout);
	}

    return 0;
}

